runtime/lua/vim/diagnostic.lua

```diff
@@ -372,43 +372,46 @@ local function to_severity(severity)
 end
 
 --- @param severity vim.diagnostic.SeverityFilter
---- @param diagnostics vim.Diagnostic[]
---- @return vim.Diagnostic[]
-local function filter_by_severity(severity, diagnostics)
-  if not severity then
-    return diagnostics
-  end
-
+--- @return fun(vim.Diagnostic):boolean
+local function severity_predicate(severity)
   if type(severity) ~= 'table' then
     severity = assert(to_severity(severity))
-    --- @param t vim.Diagnostic
-    return vim.tbl_filter(function(t)
-      return t.severity == severity
-    end, diagnostics)
+    ---@param d vim.Diagnostic
+    return function(d)
+      return d.severity == severity
+    end
   end
-
   if severity.min or severity.max then
     --- @cast severity {min:vim.diagnostic.Severity,max:vim.diagnostic.Severity}
     local min_severity = to_severity(severity.min) or M.severity.HINT
     local max_severity = to_severity(severity.max) or M.severity.ERROR
 
-    --- @param t vim.Diagnostic
-    return vim.tbl_filter(function(t)
-      return t.severity <= min_severity and t.severity >= max_severity
-    end, diagnostics)
+    --- @param d vim.Diagnostic
+    return function(d)
+      return d.severity <= min_severity and d.severity >= max_severity
+    end
   end
 
   --- @cast severity vim.diagnostic.Severity[]
-
   local severities = {} --- @type table<vim.diagnostic.Severity,true>
   for _, s in ipairs(severity) do
     severities[assert(to_severity(s))] = true
   end
 
   --- @param t vim.Diagnostic
-  return vim.tbl_filter(function(t)
+  return function(t)
```

Comment:
Any reason to stick with `t` here instead of `d` like the others above?

Reply:
No real reason, I just missed it